sql/stats/join_test.go

```diff
@@ -0,0 +1,449 @@
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{0}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{10}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{30}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{50}, BoundCnt: 1},
+			},
+			right: []sql.HistogramBucket{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{30}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{50}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{60}, BoundCnt: 1},
+			},
+			expLeft: []sql.HistogramBucket{
+				&Bucket{RowCnt: 30, DistinctCnt: 30, BoundVal: sql.Row{30}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{50}, BoundCnt: 1},
+			},
+			expRight: []sql.HistogramBucket{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{30}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{50}, BoundCnt: 1},
+			},
+		},
+		{
+			left: []sql.HistogramBucket{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{0}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{10}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{20}, BoundCnt: 1},
+			},
+			right: []sql.HistogramBucket{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{20}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{30}, BoundCnt: 1},
+			},
+			expLeft: []sql.HistogramBucket{
+				&Bucket{RowCnt: 30, DistinctCnt: 30, BoundVal: sql.Row{20}, BoundCnt: 1},
+			},
+			expRight: []sql.HistogramBucket{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{20}, BoundCnt: 1},
+			},
+		},
+		{
+			left: []sql.HistogramBucket{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{0}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{10}, BoundCnt: 1},
+			},
+			right: []sql.HistogramBucket{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{20}, BoundCnt: 1},
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{30}, BoundCnt: 1},
+			},
+			expLeft: []sql.HistogramBucket{
+				&Bucket{RowCnt: 20, DistinctCnt: 20, BoundVal: sql.Row{10}, BoundCnt: 1},
+			},
+			expRight: []sql.HistogramBucket{
+				&Bucket{RowCnt: 10, DistinctCnt: 10, BoundVal: sql.Row{20}, BoundCnt: 1},
+			},
+		},
+	}
+
+	cmp := func(i, j sql.Row) (int, error) {
+		return types.Int64.Compare(i[0], j[0])
+	}
+
+	for i, tt := range tests {
+		t.Run(fmt.Sprintf("alignment test %d", i), func(t *testing.T) {
+			lCmp, rCmp, err := AlignBuckets(tt.left, tt.right, []sql.Type{types.Int64}, []sql.Type{types.Int64}, cmp)
+			require.NoError(t, err)
+			compareHist(t, tt.expLeft, lCmp)
+			compareHist(t, tt.expRight, rCmp)
+		})
+	}
+}
+
+func TestJoin(t *testing.T) {
+	tests := []struct {
+		left  sql.Histogram
+		right sql.Histogram
+		exp   sql.Histogram
+	}{
+		{
+			left: []sql.HistogramBucket{
```

Comment:
I want to make sure I'm understanding this example right.

On the left side of the merge, there are 20 distinct values, assumed to be uniformly distributed between 0 and 10.

On the right side of the merge, there are 10 distinct values, assumed to be uniformly distributed between 0 and 20.

We expect the histogram for the merge to say that there are 10 distinct values uniformly distributed between 0 and 10.

This doesn't square with my intuition, since I would expect only half of the values from the right would be possible join candidates, so the expected number of rows should be at most 5.

I may not be properly understanding what the histogram represents or how it's being computed though.

Reply:
I think what you are suggesting is that we could truncate the RHS to a bound value of 10, and delete half of the contents. But the valid ranges are negative infinity to bound value. Most of the keys might be negative, all of the keys could overlap, or none could. Different databases handle the first bucket special case differently, I haven't had a lot of time to test alternatives.